### PR TITLE
ENH:optimize: Rewrite nnls in Python

### PR DESCRIPTION
**Pull request by ilayn**
#### Reference issue
Part of #18566, as one of the low-hanging fruits.

#### What does this implement/fix?
 - Pythonized Nonnegative Least Squares function optimize.nnls
 - Implemented the faster NNLS instead of the classical NNLS algorithm which is typically very slow. For large problems the performance difference is very obvious.
 - Added a tolerance keyword `atol` to control the inner iteration behavior. Since the algorithm errors out, the user needs something to relax the problem. I used `atol` just in case in the future we need `rtol`, so we don't need to deprecate `tol` keyword. 
 - Added two more tests and that's 100% improvement on the number of tests. 
 - Deleted the fortran77 code. 

Since I am not super up-to-date with `optimize` etiquette, please let me know if I'm going against the grain. 

This PR now opens up the possibility to solve multiple RHS through vBenthem, Keenan,  https://doi.org/10.1002/cem.889 but since there are too many things happening, I'd like to leave it to later as an enhancement. Also now code is a bit more readable (or at least familiar syntax) so probably if needed can be Cythonizable.

### CODE REVIEW

File: scipy/optimize/_nnls.py [hunk 1]
```diff
@@ -1,10 +1,11 @@
...
-from numpy import asarray_chkfinite, zeros, double
+import numpy as np
+from scipy.linalg import solve
+
 
 __all__ = ['nnls']
 
 
-def nnls(A, b, maxiter=None):
+def nnls(A, b, maxiter=None, atol=None):
```
Comment (j-bowhay) [inline scipy/optimize/_nnls.py:8]: Maybe we should consider?
```suggestion
def nnls(A, b, maxiter=None, *, atol=None):
```

File: scipy/optimize/_nnls.py [hunk 2]
```diff
@@ -1,10 +1,11 @@
...
     """
     Solve ``argmin_x || Ax - b ||_2`` for ``x>=0``. This is a wrapper
     for a FORTRAN non-negative least squares solver.
```
Comment (j-bowhay) [inline scipy/optimize/_nnls.py:11]:
```suggestion
    Solve ``argmin_x || Ax - b ||_2`` for ``x>=0``.
```
:smile:

Comment (ilayn) [reply scipy/optimize/_nnls.py:11]: Dang, Fortran is figthing back...

File: scipy/optimize/_nnls.py (hunk 1 above)
Comment (ilayn) [reply scipy/optimize/_nnls.py:8]: I think you are right, we should probably do this more often.

**Comment by dschmitz89** [comment]
Good work here! One question from me: Are there benchmarks for small matrices? `nnls` can be the bottleneck in nested optimization problems. I am not sure if a pure numpy implementation is as fast there as the classic fortran code. For larger problems I expect the new implementation to be faster though due to the algorithmic improvements.

**Comment by ilayn** [comment]
Ah good one, let me run some benchmarks

File: scipy/optimize/_nnls.py [hunk 3]
```diff
@@ -72,14 +88,73 @@ def nnls(A, b, maxiter=None):
...
+
+
+def _nnls(A, b, maxiter=None, tol=None):
+    """
+    This is a single RHS algorithm from ref [2] above. For multiple RHS
+    support, the algorithm is given in  :doi:`10.1002/cem.889`
+    """
+    m, n = A.shape
+
+    AtA = A.T @ A
```
Comment (ev-br) [inline scipy/optimize/_nnls.py:105]: Condition number? Looks like an SVD use case.

Comment (ilayn) [reply scipy/optimize/_nnls.py:105]: Didn't understand your comment. Where do you want to use SVD?

**Comment by ilayn** [comment]
image

For small arrays indeed there is some performance loss with this PR. The quickest way is to cythonize it obviously. But not sure if it is worth it since I don't know how much this is being used..

File: scipy/optimize/_nnls.py (hunk 3 above)
Comment (ev-br) [reply scipy/optimize/_nnls.py:105]: You are constructing `A.T @ A`, which has the condition number a square of the condition number of `A`. And you're using it as a l.h.s. of a linear solve routine.

**Comment by ev-br** [comment]
Also it'd be useful IMO to compare to https://github.com/scipy/scipy/blob/v1.10.1/scipy/optimize/_lsq/bvls.py, which is user-facing through `lsq_linear`.

File: scipy/optimize/_nnls.py (hunk 3 above)
Comment (ilayn) [reply scipy/optimize/_nnls.py:105]: True, albeit partial solves. The algorithm goes through submatrices of `AtA` to only deal with active constraints hence the `A[P, P]` dance.

**Comment by dschmitz89** [comment]
> image
> 
> For small arrays indeed there is some performance loss with this PR. The quickest way is to cythonize it obviously. But not sure if it is worth it since I don't know how much this is being used.. 

This is impressive. For small problems there is a minor performance loss but I don't think it must be a blocker here. Cythonization can be done in a follow up.

**Comment by ilayn** [comment]
What do the optimize residents think about this rewrite? Should we go full Cython or get this in?

**Comment by andyfaff** [comment]
Sorry, not familiar with this area. Id say if the python code is clean, and there's a performance benefit, then a merge should occur as is

**Comment by ilayn** [comment]
OK let's get it in. Since we still have some time for v.1.12 we will have the chance to remedy should something arise. Thanks everyone.

**Pull request merged**